### DISCUSSION TITLE
NotImplementedError

### DISCUSSION THREAD

**Discussion by MuradSan** [ultralytics/ultralytics#8844] [Q&A]
Hello, 
I am new with the Yolo and Pytorch, sorry if this is a noob question. Apparently I am having an error when I want  to run this code:

```python
from ultralytics import YOLO
model = YOLO('yolov8m.pt')
model.train(data=r'C:\spacecraft-pose-object-detection-runtime\notebooks\train\datasets\data.yaml',epochs=100,patience=20,batch=16, device = 0)
``` 
I get this error:

> ---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[8], [line 1](vscode-notebook-cell:?execution_count=8&line=1)
----> [1](vscode-notebook-cell:?execution_count=8&line=1) model.train(data=r'C:\spacecraft-pose-object-detection-runtime\notebooks\train\datasets\data.yaml',epochs=100,patience=20,batch=16, device = 0)

File [c:\Users\murad\AppData\Local\Programs\Python\Python311\Lib\site-packages\ultralytics\engine\model.py:655](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/model.py:655), in Model.train(self, trainer, **kwargs)
    [652](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/model.py:652)             pass
    [654](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/model.py:654) self.trainer.hub_session = self.session  # attach optional HUB session
--> [655](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/model.py:655) self.trainer.train()
    [656](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/model.py:656) # Update model and cfg after training
    [657](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/model.py:657) if RANK in (-1, 0):

File [c:\Users\murad\AppData\Local\Programs\Python\Python311\Lib\site-packages\ultralytics\engine\trainer.py:213](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:213), in BaseTrainer.train(self)
    [210](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:210)         ddp_cleanup(self, str(file))
    [212](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:212) else:
--> [213](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:213)     self._do_train(world_size)

File [c:\Users\murad\AppData\Local\Programs\Python\Python311\Lib\site-packages\ultralytics\engine\trainer.py:327](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:327), in BaseTrainer._do_train(self, world_size)
    [325](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:325) if world_size > 1:
    [326](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:326)     self._setup_ddp(world_size)
--> [327](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:327) self._setup_train(world_size)
    [329](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:329) nb = len(self.train_loader)  # number of batches
    [330](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:330) nw = max(round(self.args.warmup_epochs * nb), 100) if self.args.warmup_epochs > 0 else -1  # warmup iterations

File [c:\Users\murad\AppData\Local\Programs\Python\Python311\Lib\site-packages\ultralytics\engine\trainer.py:271](file:///C:/Users/murad/AppData/Local/Programs/Python/Python311/Lib/site-packages/ultralytics/engine/trainer.py:271), in BaseTrainer._setup_train(self, world_size)
...
PythonTLSSnapshot: registered at ..\aten\src\ATen\core\PythonFallbackKernel.cpp:162 [backend fallback]
FuncTorchDynamicLayerFrontMode: registered at ..\aten\src\ATen\functorch\DynamicLayer.cpp:494 [backend fallback]
PreDispatch: registered at ..\aten\src\ATen\core\PythonFallbackKernel.cpp:166 [backend fallback]
PythonDispatcher: registered at ..\aten\src\ATen\core\PythonFallbackKernel.cpp:158 [backend fallback]

Those are the packages I have installed:
Ultralytics YOLOv8.1.26 🚀 Python-3.11.8 torch-2.2.1+cu118 CUDA:0 (NVIDIA GeForce GTX 1070 Ti, 8192MiB)
torch              2.2.1+cu118
torchaudio         2.2.1+cu118
torchvision        0.17.1
ultralytics        8.1.26

I would appreciate all the suggestions 🙏

**Comment by pderrenger** [comment]
Hello! No worries, we're here to help, and your question is a great one! 😊 It seems like you're encountering a `NotImplementedError` when trying to train a model. This error typically occurs when a specific method or operation isn't supported or hasn't been implemented yet for the version you're using.

Given the details you've provided, everything seems to be set up correctly. However, it's worth noting that the `train` method directly on the `YOLO` object from the Ultralytics package is not the standard approach for training models. Instead, training is usually handled through the command line interface (CLI) or by using the training scripts provided in the Ultralytics YOLO repository.

For training a model with a custom dataset, you might want to follow this approach:

1. **Prepare your dataset**: Ensure your `data.yaml` is correctly set up with paths to your train/val datasets and the number of classes.
2. **Use the CLI for training**: Open your terminal and navigate to the directory where your dataset and `data.yaml` are located. Then, run the following command:

```bash
yolo train data='path/to/your/data.yaml' model='yolov8m.pt' epochs=100 patience=20 batch=16 device=0
```

Make sure to replace `'path/to/your/data.yaml'` with the actual path to your `data.yaml` file.

If you're keen on using Python code for training, you might want to check out the example training scripts in the Ultralytics YOLO GitHub repository for guidance on setting up your training pipeline.

I hope this helps! If you have any more questions or run into issues, feel free to ask. Good luck with your YOLO journey! 🚀

**Reply by MuradSan** [reply]
Thanks a lot ! :)

**Reply by pderrenger** [reply]
Certainly! Could you please provide the specific GitHub issue comment you'd like me to respond to? Without the actual comment, I'm unable to craft a precise and helpful reply. Looking forward to assisting you further! 😊
